server/lib/manifest.py

```diff
@@ -0,0 +1,382 @@
+
+        external_folders_files = list()
+
+        """
+        Handle objects that are in the dataSet, ie files that point to external sources.
+        Some of these sources may be datasets from publishers. We need to save information
+        about the source so that they can added to the Datasets section.
+        """
+        external_folders_files = self.add_tale_datasets(external_folders_files)
+
+        # Add records for the remote files that exist under a folder
+        for folder_record in external_folders_files:
+            if folder_record['file_iterator'] is None:
+                continue
+            for file_record in folder_record['file_iterator']:
+                # Check if the file points to an external resource
+                if 'linkUrl' in file_record[1]:
+                    bundle = self.create_bundle('../data/' + os.path.dirname(file_record[0]),
+                                                file_record[1]['name'])
+                    record = self.create_aggregation_record(file_record[1]['linkUrl'],
+                                                            bundle,
+                                                            folder_record.get('dataset_identifier'))
+                    self.manifest['aggregates'].append(record)
+
+    def add_tale_datasets(self, folder_files):
+        """
+        Adds information about the contents of `dataSet` to the manifest
+
+        :param folder_files: A list of objects that represent externally defined data
+        """
+        for obj in self.tale['dataSet']:
+            if obj['_modelType'] == 'folder':
+                folder = self.folderModel.load(obj['itemId'],
+                                               user=self.user,
+                                               level=AccessType.READ)
+                if folder:
+                    # Check if it's a dataset by checking for meta.identifier
+                    folder_meta = folder.get('meta')
+                    if folder_meta:
+                        dataset_identifier = folder_meta.get('identifier')
+                        if dataset_identifier:
+                            self.datasets.add(obj['itemId'])
+                            folder_files.append(
+                                {"dataset_identifier": dataset_identifier,
+                                 "provider": folder_meta.get('provider'),
+                                 "file_iterator": self.folderModel.fileList(folder,
+                                                                            user=self.user,
+                                                                            data=False)
+                                 })
+
+                    else:
+                        folder_files.append({"file_iterator": self.folderModel.fileList(
+                            folder,
+                            user=self.user,
+                            data=False)})
```

Comment:
In case of `folder` I'd do something like this:
```
if obj['_modelType'] == 'folder':
    try:
        folder = self.folderModel.load(obj['itemId'],
                                       user=self.user,
                                       level=AccessType.READ,
                                       exc=True)
        # KK: Looks like we always need this:
        folder_files_obj = { 
            'file_iterator': self.folderModel.fileList(folder, user=self.user, data=False)
        }

        try:
            folder_files_obj['dataset_identifier'] = folder['meta']['identifier']
            folder_files_obj['provider'] = folder['meta']['provider']
            self.datasets.add(obj['itemId'])
        except (KeyError, TypeError):
            # KK: is there a chance we can have one, but not the other?
            # in that case maybe it should be split.
            pass

        folder_files.append(folder_files_obj)
    except ValidationException:
        pass  # KK: No such folder
```

Reply:
Made this change, and mirrored it in the section below where folderless items in dataSet are handled.